core/src/ipc.rs

```diff
@@ -131,3 +136,105 @@ impl<T> FromRawFd for IpcReceiver<T> {
         }
     }
 }
+
+/// Creates a pair of sockets that are meant for passing file descriptors to partitions.
+pub fn io_pair<T>() -> TypedResult<(IoSender<T>, IoReceiver<T>)>
+where
+    T: AsRawFd,
+{
+    let (tx, rx) = socketpair(
+        AddressFamily::Unix,
+        SockType::Datagram,
+        None,
+        SockFlag::empty(),
+    )
+    .typ(SystemError::Panic)?;
+    unsafe { Ok((IoSender::from_raw_fd(tx), IoReceiver::from_raw_fd(rx))) }
+}
+
+#[derive(Debug)]
+/// Internal data type for the IO resource sender
+pub struct IoSender<T> {
+    socket: UnixDatagram,
+    _p: PhantomData<T>,
+}
+
+impl<T> IoSender<T>
+where
+    T: AsRawFd,
+{
+    /// Sends a resource to the receiving socket.
+    pub fn try_send(&self, resource: T) -> TypedResult<()> {
+        let fds = [resource.as_raw_fd()];
+        let cmsg = [ControlMessage::ScmRights(&fds)];
+        // Have to send some bytes along with the ancillary data.
```

Comment:
Have you tried sending an empty `IoSlice` along?

Reply:
I don't remember. But unx(7) states

> At least one byte of real data should be sent when sending ancillary data.  On Linux, this is required to successfully send ancillary data over a UNIX domain stream socket.